### PR TITLE
Facility to generate a blocktree prune list using ledger tool

### PR DESCRIPTION
**Pull request by pgarg66**
#### Problem
Need a way to boot a node from an existing ledger, but till a certain block height.

#### Summary of Changes
Added an option to ledger tool to find rooted slots and generate a list of available slots/hashes.

### CODE REVIEW

File: core/src/blocktree_processor.rs [hunk 1]
```diff
@@ -205,12 +220,14 @@ pub fn process_blocktree(
         }
 
         if !entries.is_empty() {
-            if !entries.verify(&last_entry_hash) {
-                warn!(
-                    "Ledger proof of history failed at slot: {}, entry: {}",
-                    slot, entry_height
-                );
-                return Err(BlocktreeProcessorError::LedgerVerificationFailed);
+            if rollback_height.is_none() {
```
Comment (mvines) [inline core/src/blocktree_processor.rs]: Can you please make skipping the verify a separate option to `process_blocktree...`, I'll have use for it soon outside of ledger rollback  :)

File: ledger-tool/src/main.rs [hunk 2]
```diff
@@ -58,6 +61,7 @@ fn output_ledger(blocktree: Blocktree, starting_slot: u64, method: LedgerOutputM
 }
 
 fn main() {
+    const DEFAULT_ROOT_COUNT: &str = "32";
```
Comment (mvines) [inline ledger-tool/src/main.rs]: default to 1 please

File: ledger-tool/src/main.rs [hunk 3]
```diff
@@ -82,6 +86,29 @@ fn main() {
         .subcommand(SubCommand::with_name("print").about("Print the ledger"))
         .subcommand(SubCommand::with_name("json").about("Print the ledger in JSON format"))
         .subcommand(SubCommand::with_name("verify").about("Verify the ledger's PoH"))
+        .subcommand(SubCommand::with_name("prune").about("Prune the ledger at the block height").arg(
+            Arg::with_name("valid_slot_list")
```
Comment (mvines) [inline ledger-tool/src/main.rs]: nit: just remove "valid" from slot_list/slot-list and the `.help()`.  I feel "valid" is implicit, why give an "invalid" list :)

File: ledger-tool/src/main.rs [hunk 4]
```diff
@@ -82,6 +86,29 @@ fn main() {
...
+                .long("valid-slot-list")
+                .value_name("FILENAME")
+                .takes_value(true)
+                .help("The location of the YAML file with a list of valid rollback slot heights and hashes"),
+        ))
+        .subcommand(SubCommand::with_name("list-roots").about("Output upto last <num-roots> root hashes and their heights starting at the given block height").arg(
```
Comment (mvines) [inline ledger-tool/src/main.rs]: How about:
1) Add a `--slot-list <FILE>` arg, to write the roots out to a file that can be easily provide to `prune`
2) Rename `list-roots` command to `find-roots`?

File: ledger-tool/src/main.rs [hunk 5]
```diff
@@ -123,6 +150,84 @@ fn main() {
...
+                    .last()
+                    .expect("Failed to find a valid slot");
+                println!("Prune at slot {:?} hash {:?}", target_slot, target_hash);
+            }
+        }
+        ("list-roots", Some(args_matches)) => {
+            let max_height = if let Some(height) = args_matches.value_of("max_height") {
+                usize::from_str(height).expect("Maximum height must be a number")
+            } else {
+                panic!("Maximum height must be provided");
```
Comment (rob-solana) [inline ledger-tool/src/main.rs]: better to make clap require this parameter to the command, I think.

Comment (pgarg66) [reply ledger-tool/src/main.rs]: It is a required parameter in clap. Line 101

**Pull request merged**